### ISSUE TITLE
This crate isn't unsound on its own, but causes UB in practice 

### ISSUE THREAD

**Issue by Lokathor** [Serentty/wide-literals#1]
A reference type restricts the span of memory it allows to be *only* it's actual span, without allowing access "past the first element" like a pointer does. This means that having the macro create `&'static u16` will allow you to legally access *only* that one specific u16, not the others after it. So in practice this crate is unsound when you're using it and then calling functions that expect to be able to read until a 0.

**Comment by Serentty** [comment]
Oh wow, I had honestly forgotten that I had uploaded this. I think I mostly did so so that I could easily include it in my other projects.

Would it be fine if I had the macro create a raw pointer instead?

**Comment by Lokathor** [comment]
yeah, using the .as_ptr() method is probably the best solution (but it is a breaking change unfortunately).

**Comment by Serentty** [comment]
I would be pretty shocked if anyone is using this crate, but if they are, sorry to them about that.

**Comment by Lokathor** [comment]
I wouldn't have known about it, except that you got used as an example in support of an RFC, and I just happened to open your docs and see that something looked fishy.

**Comment by Serentty** [comment]
Oh wow, where was the thread?

**Comment by Lokathor** [comment]
https://github.com/rust-lang/rfcs/pull/3267

This crate was given as an example of a good fit for the RFC's proposal in the first few comments.

However, I don't think the RFC is a good idea.

**Comment by ChayimFriedman2** [comment]
I _do_ think the RFC is a good idea, but this is not surely UB: [Storing an object as &Header, but reading the data past the end of the header](https://github.com/rust-lang/unsafe-code-guidelines/issues/256).

This does need to be fixed, but maybe if the breakage is too high, we should not do it unless it is decided to be UB?

Anyway, I think it is good to bring this example in the abovementioned UCG issue.

**Comment by Lokathor** [comment]
I haven't read that thread in a while so i don't know for sure, but the parts i skimmed again just now always had Ralf saying "this is UB", so which reply in that thread made this not be UB?

**Comment by ChayimFriedman2** [comment]
Neither have I, but from what I remember Ralf indeed wants to make that UB but there are people that don't (mainly because they use it, or depends on crate that use it) and so the question is what we decide.

Of course, like everything undecided about UB in Rust you should avoid it, and if enough people will maybe we will be able to finally declare it as UB :)

**Comment by Serentty** [comment]
So it’s not clear yet whether or not this is undefined behaviour? That’s a tricky one. I will try to change the crate then. I have been procrastinating. It’s a really simple change.

**Comment by Lokathor** [comment]
My understanding is that it *currently is UB* and Ralf wants to *keep it as UB*.

**Comment by CAD97** [comment]
AIUI Ralf is optimistic that the next iteration of the SB model will not *require* strict bounds/subobject slicing for references, and that choosing whether subobject slicing is required by the borrow model will be an orthogonal choice to the soundness of the borrow model and the main optimizations it justifies.

In fact, this is IIUC necessary for references to `extern type` to function soundly within the borrow model.

I would not be surprised if subobject slicing ends up on the tower of weakenings (i.e. Miri, CHERI, etc. check it, but it's fine on e.g. x86_64) but I also wouldn't be surprised if it's required. Best practice is and probably always will be to play nice in the face of subobject slicing.

https://github.com/rust-lang/unsafe-code-guidelines/issues/256#issuecomment-1044915400

(Hi, sorry, just stumbled into this while pulling references on the &Header discussion for someone else)

**Comment by Serentty** [comment]
I keep forgetting to get around to this.
